test/unittest/unittest_properties.cpp

```diff
@@ -0,0 +1,134 @@
+// SPDX-License-Identifier: Apache-2.0
+/**
+ * Copyright (C) 2021 Jihoon Lee <jhoon.it.lee@samsung.com>
+ *
+ * @file unittest_properties.h
+ * @date 09 April 2021
+ * @brief This file contains test and sepecification of properties
+ * @see	https://github.com/nnstreamer/nntrainer
+ * @author Jihoon Lee <jhoon.it.lee@samsung.com>
+ * @bug No known bugs except for NYI items
+ */
+#include <gtest/gtest.h>
+
+#include <base_properties.h>
+#include <util_func.h>
+
+namespace { /**< define a property for testing */
+
+struct banana_prop_tag : nntrainer::int_prop_tag {};
+
+class NumBanana : public nntrainer::Property<int> {
+public:
+  NumBanana() { set(1); }                          /**< default value if any */
+  static constexpr const char *key = "num_banana"; /**< unique key to access */
+  using prop_tag = banana_prop_tag;                /**< property type */
+
+  bool is_valid(const int &v) override { return v >= 0; }
+};
+
+class QualityOfBanana : public nntrainer::Property<std::string> {
+public:
+  static constexpr const char *key = "quality_banana";
+  using prop_tag = nntrainer::str_prop_tag;
+
+  bool is_valid(const std::string &v) override {
+    /// assuming quality of banana property must ends with word "good";
+    return nntrainer::endswith(v, "good");
+  }
+};
+} // namespace
+
+TEST(BasicProperty, tagCast) {
+  EXPECT_EQ(1, 1); /**< this is to prevent no assert tc from TCM */
+
+  { /**< tag_cast simple cast */
+    using T =
+      nntrainer::tag_cast<banana_prop_tag, nntrainer::int_prop_tag>::type;
+    ::testing::StaticAssertTypeEq<T, nntrainer::int_prop_tag>();
+  }
+
+  { /**< tag_cast ranged cast */
+    using T = nntrainer::tag_cast<banana_prop_tag, nntrainer::vector_prop_tag,
+                                  nntrainer::int_prop_tag>::type;
+    ::testing::StaticAssertTypeEq<T, nntrainer::int_prop_tag>();
+  }
+
+  { /**< tag_cast does not have appropriate candidates */
+    using T = nntrainer::tag_cast<banana_prop_tag, int, std::string>::type;
+    ::testing::StaticAssertTypeEq<T, banana_prop_tag>();
+  }
+}
+
+TEST(BasicProperty, valid_p) {
+  { /** set -> get / to_string */
+    NumBanana b;
```

Comment:
does this mean that every custom layer when they want new properties will have to make new class like numBanana or qualityOfBanana?

Reply:
If it is common, layer dose not need to make a new class else yes.

Reply:
So, all common properties must be opened to devel.